test/integration/http_timeout_integration_test.cc

```diff
@@ -47,6 +47,53 @@ TEST_P(HttpTimeoutIntegrationTest, GlobalTimeout) {
   EXPECT_EQ("504", response->headers().Status()->value().getStringView());
 }
 
+// Regression test for https://github.com/envoyproxy/envoy/issues/7154 in which
+// resetStream() was only called after a response timeout for upstream requests
+// that had not received headers yet. This meant that decodeData might be
+// called on a destroyed UpstreamRequest.
+TEST_P(HttpTimeoutIntegrationTest, GlobalTimeoutAfterHeadersBeforeBodyResetsUpstream) {
+  initialize();
+
+  codec_client_ = makeHttpConnection(makeClientConnection(lookupPort("http")));
+  Http::TestHeaderMapImpl request_headers{{":method", "POST"},
+                                          {":path", "/test/long/url"},
+                                          {":scheme", "http"},
+                                          {":authority", "host"},
+                                          {"x-forwarded-for", "10.0.0.1"},
+                                          {"x-envoy-upstream-rq-timeout-ms", "100"}};
+  auto encoder_decoder = codec_client_->startRequest(request_headers);
```

Comment:
use `auto` only if the type is obvious (e.g. cast, make_unique/shared, assigning literals).